core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -208,6 +209,10 @@ class TransactionCoordinator(brokerId: Int,
     if (transactionalId == null || transactionalId.isEmpty) {
       responseCallback(Errors.INVALID_REQUEST)
     } else {
+      // if there is any partitions unknown in the metadata cache, return immediately to client
+      if (partitions.exists(tp => !metadataCache.contains(tp.topic, tp.partition)))
+        responseCallback(Errors.UNKNOWN_TOPIC_OR_PARTITION)
```

Comment:
> If we want to keep this behavior, I think we may want to fix KAFKA-5251 as well.

I think KAFKA-5251 is an optimization, while we want to fix the client behavior that if the topic partition gets deleted while the producer is having an ongoing transaction, it should be able to detect this and proceed instead of retrying, right?